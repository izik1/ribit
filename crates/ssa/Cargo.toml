[package]
authors.workspace = true
edition.workspace = true
name = "ribit-ssa"
publish.workspace = true
version.workspace = true

[dependencies]
ribit-core.workspace = true

[dev-dependencies]
expect-test.workspace = true
ribit-asm.workspace = true
