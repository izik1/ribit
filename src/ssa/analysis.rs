use crate::ssa::{Id, Instruction, Source, StackIndex};
use std::collections::HashMap;
use std::convert::TryFrom;
use std::fmt;

fn lifetime_instruction<F: FnMut(&mut Lifetimes, Source, usize)>(
    instr: &Instruction,
    idx: usize,
    lifetimes: &mut Lifetimes,
    mut update: F,
) {
    match instr {
        Instruction::Fence => {}

        Instruction::BinOp {
            dest, src1, src2, ..
        }
        | Instruction::Cmp {
            dest, src1, src2, ..
        } => {
            update(lifetimes, Source::Id(*dest), idx);
            update(lifetimes, *src1, idx);
            update(lifetimes, *src2, idx);
        }

        Instruction::ReadReg { dest, base, .. } => {
            update(lifetimes, *base, idx);
            update(lifetimes, Source::Id(*dest), idx);
        }

        Instruction::Arg { dest, .. }
        | Instruction::LoadConst { dest, .. }
        | Instruction::ReadStack { dest, .. } => {
            update(lifetimes, Source::Id(*dest), idx);
        }

        Instruction::WriteStack { dest: _, src } => {
            update(lifetimes, Source::Id(*src), idx);
        }

        Instruction::WriteReg { src, base, .. } => {
            update(lifetimes, *base, idx);
            update(lifetimes, *src, idx);
        }

        Instruction::ReadMem {
            dest, src, base, ..
        } => {
            update(lifetimes, *base, idx);
            update(lifetimes, Source::Id(*dest), idx);
            update(lifetimes, *src, idx);
        }

        Instruction::WriteMem {
            addr, src, base, ..
        } => {
            update(lifetimes, *base, idx);
            update(lifetimes, *addr, idx);
            update(lifetimes, *src, idx);
        }

        Instruction::Select {
            dest,
            cond,
            if_true,
            if_false,
        } => {
            update(lifetimes, Source::Id(*dest), idx);
            update(lifetimes, *cond, idx);
            update(lifetimes, *if_true, idx);
            update(lifetimes, *if_false, idx);
        }

        Instruction::Ret { addr, code } => {
            update(lifetimes, *addr, idx);
            update(lifetimes, *code, idx);
        }
    }
}

#[derive(Copy, Clone, Eq, PartialEq)]
pub struct Lifetime {
    pub start: usize,
    pub end: usize,
}

impl Lifetime {
    #[must_use]
    pub fn is_alive_after(self, idx: usize) -> bool {
        idx >= self.start && idx <= self.end
    }

    #[must_use]
    pub fn is_empty(self) -> bool {
        self.start == self.end
    }
}

pub type Lifetimes = HashMap<Id, Lifetime>;

fn update_lifetime(lifetimes: &mut Lifetimes, src: Source, idx: usize) {
    if let Some(id) = src.id() {
        lifetimes
            .entry(id)
            .and_modify(|it| it.end = idx)
            .or_insert(Lifetime {
                start: idx,
                end: idx,
            });
    }
}

#[must_use]
pub fn lifetimes(instrs: &[Instruction]) -> Lifetimes {
    let mut lifetimes = HashMap::with_capacity(instrs.len());

    for (idx, instr) in instrs.iter().enumerate() {
        lifetime_instruction(instr, idx, &mut lifetimes, update_lifetime);
    }

    lifetimes
}

#[must_use]
pub fn surrounding_usages(instrs: &[Instruction], needle: usize) -> Lifetimes {
    let mut lifetimes = HashMap::new();

    for (idx, instr) in instrs.iter().enumerate().take(needle) {
        lifetime_instruction(instr, idx, &mut lifetimes, |lifetimes, src, idx| {
            if let Some(id) = src.id() {
                lifetimes.insert(
                    id,
                    Lifetime {
                        start: idx,
                        end: idx,
                    },
                );
            }
        })
    }

    for (idx, instr) in instrs.iter().enumerate().skip(needle) {
        lifetime_instruction(instr, idx, &mut lifetimes, |lifetimes, src, idx| {
            if let Some(id) = src.id() {
                lifetimes.entry(id).and_modify(|it| {
                    if it.is_empty() {
                        it.end = idx
                    }
                });
            }
        })
    }

    lifetimes
}

// todo: replace this with a Vec<...> `rather` than the hashmap,
//  this is because StackIndexes should always be _used_ linearly.
#[must_use]
pub fn stack_lifetimes(graph: &[Instruction]) -> HashMap<StackIndex, Vec<(usize, usize)>> {
    let mut lifetimes = HashMap::new();
    for (idx, instr) in graph.iter().enumerate() {
        match instr {
            Instruction::WriteStack { dest, src: _ } => lifetimes
                .entry(*dest)
                .or_insert_with(Vec::new)
                .push((idx, idx)),
            Instruction::ReadStack { dest: _, src } => {
                lifetimes.get_mut(src).unwrap().last_mut().unwrap().1 = idx;
            }
            _ => {}
        }
    }

    lifetimes
}

/// Get the highest used `StackIndex` for a given graph.
#[must_use]
pub fn max_stack(graph: &[Instruction]) -> Option<StackIndex> {
    let mut max: Option<StackIndex> = None;

    for instr in graph {
        match instr {
            Instruction::WriteStack {
                dest: stack_idx,
                src: _,
            }
            | Instruction::ReadStack {
                dest: _,
                src: stack_idx,
            } => {
                max = std::cmp::max(max, Some(*stack_idx));
            }

            _ => {}
        }
    }

    max
}

#[must_use]
pub fn min_stack(
    lifetime: Lifetime,
    stack_lts: &HashMap<StackIndex, Vec<(usize, usize)>>,
) -> StackIndex {
    stack_lts
        .iter()
        .filter_map(|(idx, stack_lts)| {
            (stack_lts
                .iter()
                .all(|stack_lt| stack_lt.1 < lifetime.start || stack_lt.0 > lifetime.end))
            .then_some(*idx)
        })
        .min()
        .unwrap_or_else(|| {
            StackIndex(u8::try_from(stack_lts.len()).expect("Ran out of stack indexes"))
        })
}

pub struct ShowLifetimes<'a, 'b> {
    lifetimes: &'a Lifetimes,
    instrs: &'b [Instruction],
}

impl<'a, 'b> ShowLifetimes<'a, 'b> {
    pub fn new(lifetimes: &'a Lifetimes, instrs: &'b [Instruction]) -> Self {
        Self { lifetimes, instrs }
    }
}

fn show_lifetime(f: &mut fmt::Formatter, lifetime: &Lifetime, idx: usize) -> fmt::Result {
    let width = (lifetime.end - lifetime.start) / 10 + 1;
    if lifetime.is_alive_after(idx) {
        write!(f, "{:1$}", lifetime.end - idx, width)
    } else {
        f.write_str(&"-".repeat(width))
    }
}

impl fmt::Display for ShowLifetimes<'_, '_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for (idx, instr) in self.instrs.iter().enumerate() {
            let mut lifetimes: Vec<Lifetime> = self.lifetimes.values().cloned().collect();
            lifetimes.sort_by(|a, b| a.start.cmp(&b.start));
            let mut lifetimes = lifetimes.iter();

            write!(f, "[")?;

            if let Some(lifetime) = lifetimes.next() {
                show_lifetime(f, lifetime, idx)?;
            }

            for lifetime in lifetimes {
                f.write_str(", ")?;
                show_lifetime(f, lifetime, idx)?;
            }

            f.write_str("] ")?;

            writeln!(f, "{}", instr)?;
        }

        Ok(())
    }
}

#[cfg(test)]
mod test {
    use super::ShowLifetimes;
    use crate::ssa::opt;

    use insta::assert_display_snapshot;

    #[test]
    fn max_lifetimes() {
        let (mut instrs, _) = crate::ssa::max_fn();

        opt::fold_and_prop_consts(&mut instrs);
        let instrs = opt::dead_instruction_elimination(&instrs);
        let instrs = opt::register_writeback_shrinking(&instrs);

        let lifetimes = super::lifetimes(&instrs);

        assert_display_snapshot!(ShowLifetimes::new(&lifetimes, &instrs));
    }
}
