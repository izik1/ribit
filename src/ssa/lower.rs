use super::{BinOp, CmpKind, Id, Instruction, Source};
use crate::ssa::eval;
use crate::{instruction, opcode, register, Width};

pub struct Context {
    next_id: Id,
    pub pc: Source,
    registers: [Option<Source>; 32],
    registers_written: u32,
    instructions: Vec<Instruction>,
}

impl Context {
    pub fn new(start_pc: u32) -> Self {
        Self {
            next_id: Id(0),
            pc: Source::Val(start_pc),
            registers: [None; 32],
            registers_written: 0x0000_0000,
            instructions: vec![],
        }
    }

    pub fn add_pc(&mut self, src: Source) -> Source {
        self.pc = self.add(src, self.pc);
        self.pc
    }

    fn new_id(&mut self) -> Id {
        let id_num = self.next_id.0;

        assert!(id_num < u16::max_value());

        std::mem::replace(&mut self.next_id, Id(id_num + 1))
    }

    fn instruction<F: FnOnce(Id) -> Instruction>(&mut self, f: F) -> Source {
        let id = self.new_id();
        self.instructions.push(f(id));
        Source::Id(id)
    }

    pub fn binop(&mut self, op: BinOp, src1: Source, src2: Source) -> Source {
        if let (Some(src1), Some(src2)) = (src1.val(), src2.val()) {
            Source::Val(eval::binop(src1, src2, op))
        } else {
            self.instruction(|dest| Instruction::BinOp {
                dest,
                src1,
                src2,
                op,
            })
        }
    }

    pub fn read_register(&mut self, reg: register::RiscV) -> Source {
        self.registers[reg.get() as usize].unwrap_or_else(|| {
            let id = self.new_id();

            self.instructions
                .push(Instruction::ReadReg { dest: id, src: reg });

            self.registers[reg.get() as usize] = Some(Source::Id(id));

            Source::Id(id)
        })
    }

    pub fn load_register(&mut self, reg: Option<register::RiscV>) -> Source {
        reg.map_or(Source::Val(0), |reg| self.read_register(reg))
    }

    pub fn read_memory(&mut self, src: Source, width: Width, sign_extend: bool) -> Source {
        self.instruction(|dest| Instruction::ReadMem {
            dest,
            src,
            width,
            sign_extend,
        })
    }

    pub fn write_register(&mut self, reg: register::RiscV, val: Source) {
        self.registers[reg.get() as usize] = Some(val);
        self.registers_written |= 1 << reg.get();
    }

    pub fn write_memory(&mut self, addr: Source, val: Source, width: Width) {
        self.instructions.push(Instruction::WriteMem {
            addr,
            src: val,
            width,
        });
    }

    pub fn or(&mut self, src1: Source, src2: Source) -> Source {
        self.binop(BinOp::Or, src1, src2)
    }

    pub fn sll(&mut self, src1: Source, src2: Source) -> Source {
        self.binop(BinOp::Sll, src1, src2)
    }

    pub fn srl(&mut self, src1: Source, src2: Source) -> Source {
        self.binop(BinOp::Srl, src1, src2)
    }

    pub fn sra(&mut self, src1: Source, src2: Source) -> Source {
        self.binop(BinOp::Sra, src1, src2)
    }

    pub fn xor(&mut self, src1: Source, src2: Source) -> Source {
        self.binop(BinOp::Xor, src1, src2)
    }

    pub fn and(&mut self, src1: Source, src2: Source) -> Source {
        self.binop(BinOp::And, src1, src2)
    }

    pub fn add(&mut self, src1: Source, src2: Source) -> Source {
        self.binop(BinOp::Add, src1, src2)
    }

    pub fn sub(&mut self, src1: Source, src2: Source) -> Source {
        self.binop(BinOp::Sub, src1, src2)
    }

    pub fn cmp(&mut self, src1: Source, src2: Source, mode: CmpKind) -> Source {
        if let (Some(src1), Some(src2)) = (src1.val(), src2.val()) {
            Source::Val(eval::cmp(src1, src2, mode))
        } else {
            self.instruction(|dest| Instruction::Cmp {
                dest,
                src1,
                src2,
                kind: mode,
            })
        }
    }

    pub fn select(&mut self, cond: Source, if_true: Source, if_false: Source) -> Source {
        if let Some(res) = eval::try_select(cond.val(), if_true.val(), if_false.val()) {
            Source::Val(res)
        } else {
            self.instruction(|dest| Instruction::Select {
                dest,
                if_true,
                if_false,
                cond,
            })
        }
    }

    pub fn fence(&mut self) {
        self.instructions.push(Instruction::Fence)
    }

    #[must_use]
    pub fn ret_with_code(mut self, addr: Source, code: Source) -> Vec<Instruction> {
        for (idx, src) in self.registers.iter().enumerate().skip(1) {
            let dest = register::RiscV::with_u8(idx as u8).unwrap();

            if let Some(src) = *src {
                if (self.registers_written >> dest.get()) & 1 == 1 {
                    self.instructions.push(Instruction::WriteReg { dest, src });
                }
            }
        }

        self.instructions.push(Instruction::Ret { addr, code });

        // todo: assert well formedness here.
        self.instructions
    }

    #[must_use]
    pub fn ret_with_addr(self, addr: Source) -> Vec<Instruction> {
        self.ret_with_code(addr, Source::Val(0))
    }

    #[must_use]
    pub fn ret(self) -> Vec<Instruction> {
        let pc = self.pc;
        self.ret_with_addr(pc)
    }
}

pub fn non_terminal(ctx: &mut Context, instruction: instruction::Instruction, len: u32) {
    match instruction {
        instruction::Instruction::J(_)
        | instruction::Instruction::Sys(_)
        | instruction::Instruction::B(_)
        | instruction::Instruction::IJump(_) => panic!("Instruction was a terminal"),

        instruction::Instruction::IMem(instruction::IMem {
            opcode,
            imm,
            rs1,
            rd,
        }) => match opcode {
            // todo: finer grained fences.
            opcode::IMem::FENCE => ctx.fence(),
            opcode::IMem::LD(width) => {
                let imm = imm as i16 as u32;

                let src = ctx.load_register(rs1);
                let offset = ctx.add(src, Source::Val(imm));

                let mem = ctx.read_memory(offset, width, true);

                if let Some(rd) = rd {
                    ctx.write_register(rd, mem);
                }
            }
            // todo: deduplicate with previous
            opcode::IMem::LDU(width) => {
                let imm = imm as i16 as u32;

                let src = ctx.load_register(rs1);
                let offset = ctx.add(src, Source::Val(imm));

                let mem = ctx.read_memory(offset, width, false);

                if let Some(rd) = rd {
                    ctx.write_register(rd, mem);
                }
            }
        },

        instruction::Instruction::I(instruction::I {
            opcode,
            imm,
            rs1,
            rd,
        }) => {
            let imm = imm as i16 as u32;
            let src = ctx.load_register(rs1);

            let res = match opcode {
                opcode::I::ADDI => ctx.add(src, Source::Val(imm)),
                opcode::I::XORI => ctx.xor(src, Source::Val(imm)),
                opcode::I::ANDI => ctx.and(src, Source::Val(imm)),
                opcode::I::SLLI => ctx.sll(src, Source::Val(imm)),
                opcode::I::SRLI => ctx.srl(src, Source::Val(imm)),
                opcode::I::SRAI => ctx.sra(src, Source::Val(imm)),
                opcode::I::ORI => ctx.or(src, Source::Val(imm)),
                opcode::I::SICond(cmp_mode) => ctx.cmp(src, Source::Val(imm), cmp_mode.into()),
            };

            if let Some(rd) = rd {
                ctx.write_register(rd, res);
            }
        }

        instruction::Instruction::R(instruction::R {
            rs1,
            rs2,
            rd,
            opcode,
        }) => {
            let src1 = ctx.load_register(rs1);
            let src2 = ctx.load_register(rs2);

            let res = match opcode {
                opcode::R::ADD => ctx.add(src1, src2),
                opcode::R::SUB => ctx.sub(src1, src2),
                opcode::R::SLL => ctx.sll(src1, src2),
                opcode::R::SCond(cmp) => ctx.cmp(src1, src2, cmp.into()),
                opcode::R::XOR => ctx.xor(src1, src2),
                opcode::R::SRL => ctx.srl(src1, src2),
                opcode::R::SRA => ctx.sra(src1, src2),
                opcode::R::OR => ctx.or(src1, src2),
                opcode::R::AND => ctx.and(src1, src2),
                opcode::R::MUL => todo!(),
                opcode::R::MULH => todo!(),
                opcode::R::MULHSU => todo!(),
                opcode::R::MULHU => todo!(),
                opcode::R::DIV => todo!(),
                opcode::R::DIVU => todo!(),
                opcode::R::REM => todo!(),
                opcode::R::REMU => todo!(),
            };

            if let Some(rd) = rd {
                ctx.write_register(rd, res);
            }
        }
        instruction::Instruction::S(instruction::S {
            width,
            rs1,
            rs2,
            imm,
        }) => {
            let imm = imm as i16 as u32;

            let src1 = ctx.load_register(rs1);
            let addr = ctx.add(src1, Source::Val(imm));

            let src2 = ctx.load_register(rs2);
            ctx.write_memory(addr, src2, width);
        }
        instruction::Instruction::U(instruction::U { opcode, imm, rd }) => {
            // ensure that the immediate only uses the upper 20 bits.
            let imm = imm & 0xffff_f000;

            let res = match opcode {
                opcode::U::LUI => Source::Val(imm),
                opcode::U::AUIPC => ctx.add(ctx.pc, Source::Val(imm)),
            };

            if let Some(rd) = rd {
                ctx.write_register(rd, res);
            }
        }
    }

    ctx.add_pc(Source::Val(len));
}

#[must_use]
pub fn terminal(
    mut ctx: Context,
    instruction: instruction::Instruction,
    len: u32,
) -> Vec<self::Instruction> {
    match instruction {
        instruction::Instruction::J(instruction::J {
            opcode: opcode::J::JAL,
            rd,
            imm,
        }) => {
            // load pc + len into rd (the link register)
            if let Some(rd) = rd {
                // we specifically need to avoid modifying `pc`, since we need the
                // old value for adding the immediate.
                let next_pc = ctx.add(ctx.pc, Source::Val(len));

                ctx.write_register(rd, next_pc);
            }

            ctx.add_pc(Source::Val(imm));

            ctx.ret()
        }

        instruction::Instruction::IJump(instruction::IJump {
            opcode: opcode::IJump::JALR,
            rd,
            imm,
            rs1,
        }) => {
            let imm = imm as i16 as u32;

            // load pc + len into rd (the link register)
            if let Some(rd) = rd {
                // we specifically need to avoid modifying `pc`, since we need the
                // old value for adding the immediate.
                let next_pc = ctx.add(ctx.pc, Source::Val(len));

                ctx.write_register(rd, next_pc);
            }

            ctx.add_pc(Source::Val(imm));

            // adding 0 is the same as not adding, so don't bother with a match here.
            if let Some(src) = rs1 {
                let src = ctx.read_register(src);
                ctx.add_pc(src);
            }

            ctx.ret()
        }

        instruction::Instruction::Sys(instruction::Sys { opcode }) => {
            use crate::ReturnCode;
            let return_code = match opcode {
                opcode::RSys::EBREAK => ReturnCode::EBreak,
                opcode::RSys::ECALL => ReturnCode::ECall,
            } as u32;

            let pc = ctx.add_pc(Source::Val(len));

            ctx.ret_with_code(pc, Source::Val(return_code))
        }

        instruction::Instruction::B(instruction::B {
            imm,
            rs1,
            rs2,
            cmp_mode,
        }) => {
            let src1 = ctx.load_register(rs1);
            let src2 = ctx.load_register(rs2);

            let cmp = ctx.cmp(src1, src2, cmp_mode.into());

            let continue_pc = ctx.add(ctx.pc, Source::Val(len));

            let jump_pc = ctx.add(ctx.pc, Source::Val(imm as i16 as u32));

            let addr = ctx.select(cmp, jump_pc, continue_pc);

            ctx.ret_with_addr(addr)
        }

        _ => panic!("Instruction wasn't a terminal"),
    }
}

#[cfg(test)]
mod test {
    use super::Context;
    use crate::DisplayDeferSlice;
    use crate::{instruction, opcode, register};

    use insta::{assert_display_snapshot, assert_ron_snapshot};

    #[test]
    fn jal_basic() {
        let ctx = Context::new(0);

        let instrs = super::terminal(
            ctx,
            instruction::Instruction::J(instruction::J {
                imm: 4096,
                rd: Some(register::RiscV::X4),
                opcode: opcode::J::JAL,
            }),
            4,
        );

        assert_display_snapshot!(DisplayDeferSlice(&instrs));
    }

    #[test]
    fn sys_break() {
        let ctx = Context::new(0);

        let instrs = super::terminal(
            ctx,
            instruction::Instruction::Sys(instruction::Sys::new(opcode::RSys::EBREAK)),
            4,
        );

        assert_display_snapshot!(DisplayDeferSlice(&instrs));
    }

    #[test]
    fn addi_nop() {
        let mut ctx = Context::new(0);
        super::non_terminal(
            &mut ctx,
            instruction::Instruction::I(instruction::I::new(0, None, None, opcode::I::ADDI)),
            4,
        );

        let instrs = super::terminal(
            ctx,
            instruction::Instruction::Sys(instruction::Sys::new(opcode::RSys::EBREAK)),
            4,
        );

        assert_display_snapshot!(DisplayDeferSlice(&instrs));
    }

    #[test]
    fn branch_0_0_eq() {
        let ctx = Context::new(0);
        let instrs = super::terminal(
            ctx,
            instruction::Instruction::B(instruction::B::new(1024, None, None, opcode::Cmp::Eq)),
            4,
        );

        assert_display_snapshot!(DisplayDeferSlice(&instrs));
    }

    #[test]
    fn branch_0_x1_eq() {
        let ctx = Context::new(0);
        let instrs = super::terminal(
            ctx,
            instruction::Instruction::B(instruction::B::new(
                1024,
                None,
                Some(register::RiscV::X1),
                opcode::Cmp::Eq,
            )),
            4,
        );

        assert_display_snapshot!(DisplayDeferSlice(&instrs));
    }

    #[test]
    fn addi_no_dest() {
        let mut ctx = Context::new(0);
        super::non_terminal(
            &mut ctx,
            instruction::Instruction::I(instruction::I::new(
                50,
                Some(register::RiscV::X1),
                None,
                opcode::I::ADDI,
            )),
            4,
        );

        let instrs = super::terminal(
            ctx,
            instruction::Instruction::Sys(instruction::Sys::new(opcode::RSys::EBREAK)),
            4,
        );

        assert_display_snapshot!(DisplayDeferSlice(&instrs));
    }

    #[test]
    fn addi_no_src() {
        let mut ctx = Context::new(0);
        super::non_terminal(
            &mut ctx,
            instruction::Instruction::I(instruction::I::new(
                50,
                None,
                Some(register::RiscV::X2),
                opcode::I::ADDI,
            )),
            4,
        );

        let instrs = super::terminal(
            ctx,
            instruction::Instruction::Sys(instruction::Sys::new(opcode::RSys::EBREAK)),
            4,
        );

        assert_display_snapshot!(DisplayDeferSlice(&instrs));
    }
}
