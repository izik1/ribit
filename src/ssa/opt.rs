use crate::ssa::{eval, Id, Instruction, Source};
use std::collections::HashMap;

fn const_id_lookup(consts: &HashMap<Id, u32>, src: Source) -> Option<u32> {
    match src {
        Source::Id(id) => Some(id),
        Source::Val(_) => None,
    }
    .and_then(|it| consts.get(&it).copied())
}

fn const_prop(consts: &HashMap<Id, u32>, src: &mut Source) -> Option<u32> {
    if let Some(v) = const_id_lookup(consts, *src) {
        *src = Source::Val(v);
        Some(v)
    } else if let Source::Val(v) = src {
        Some(*v)
    } else {
        None
    }
}

pub fn fold_and_prop_consts(graph: &mut [Instruction]) {
    let mut consts = HashMap::new();
    for instruction in graph.iter_mut() {
        let (dest, val) = match instruction {
            Instruction::LoadConst { dest, src } => (*dest, *src),

            Instruction::Fence | Instruction::Arg { .. } => continue,
            Instruction::ReadReg { base, .. } => {
                const_prop(&consts, base);
                continue;
            }

            Instruction::WriteReg { src, base, .. }
            | Instruction::ReadMem { src, base, .. }
            | Instruction::WriteMem { src, base, .. } => {
                const_prop(&consts, src);
                const_prop(&consts, base);

                continue;
            }

            Instruction::BinOp {
                dest,
                src1,
                src2,
                op,
            } => {
                let src1 = const_prop(&consts, src1);
                let src2 = const_prop(&consts, src2);

                let res = match (src1, src2) {
                    (Some(src1), Some(src2)) => eval::binop(src1, src2, *op),
                    _ => continue,
                };

                (*dest, res)
            }

            Instruction::Cmp {
                dest,
                src1,
                src2,
                kind,
            } => {
                let src1 = const_prop(&consts, src1);
                let src2 = const_prop(&consts, src2);

                // note: this explicitly doesn't simplify things like
                // `cmp eq %0, %0`, that's for a different pass
                // todo: write pass for the above.

                let res = match (src1, src2) {
                    (Some(src1), Some(src2)) => eval::cmp(src1, src2, *kind),
                    _ => continue,
                };

                (*dest, res)
            }

            Instruction::Select {
                dest,
                cond,
                if_true,
                if_false,
            } => {
                let cond = const_prop(&consts, cond);
                let if_true = const_prop(&consts, if_true);
                let if_false = const_prop(&consts, if_false);

                if let Some(res) = eval::try_select(cond, if_true, if_false) {
                    (*dest, res)
                } else {
                    continue;
                }
            }

            Instruction::Ret { addr, code } => {
                const_prop(&consts, addr);
                const_prop(&consts, code);

                continue;
            }
        };

        *instruction = Instruction::LoadConst { dest, src: val };

        consts.insert(dest, val);
    }
}

fn mark_live(live_instructions: &mut [bool; 0x1_0000], src: Source) {
    if let Source::Id(id) = src {
        live_instructions[id.0 as usize] = true;
    }
}

#[must_use]
pub fn dead_instruction_elimination(graph: &[Instruction]) -> Vec<Instruction> {
    let mut live_ids = [false; 0x1_0000];
    let mut live_instruction_count = 0;

    for instruction in graph.iter().rev() {
        match instruction {
            Instruction::Ret { addr, code } => {
                mark_live(&mut live_ids, *addr);
                mark_live(&mut live_ids, *code);

                live_instruction_count += 1;
            }

            Instruction::Fence => live_instruction_count += 1,

            Instruction::ReadReg { src: _, dest, base } => {
                mark_live(&mut live_ids, *base);

                if live_ids[dest.0 as usize] {
                    live_instruction_count += 1;
                }
            }

            Instruction::LoadConst { dest, .. } | Instruction::Arg { dest, .. } => {
                if live_ids[dest.0 as usize] {
                    live_instruction_count += 1;
                }
            }

            Instruction::ReadMem {
                src,
                base,
                dest: _,
                sign_extend: _,
                width: _,
            } => {
                // hack: I'm not sure when it's safe to remove unused memory reads,
                // so assume we never can.
                live_instruction_count += 1;
                mark_live(&mut live_ids, *src);
                mark_live(&mut live_ids, *base);
            }

            Instruction::WriteReg { src, dest: _, base } => {
                live_instruction_count += 1;
                mark_live(&mut live_ids, *src);
                mark_live(&mut live_ids, *base);
            }

            Instruction::WriteMem {
                addr,
                src,
                base,
                width: _,
            } => {
                live_instruction_count += 1;
                mark_live(&mut live_ids, *addr);
                mark_live(&mut live_ids, *src);
                mark_live(&mut live_ids, *base);
            }

            Instruction::BinOp {
                dest, src1, src2, ..
            }
            | Instruction::Cmp {
                dest, src1, src2, ..
            } => {
                if live_ids[dest.0 as usize] {
                    live_instruction_count += 1;
                    mark_live(&mut live_ids, *src1);
                    mark_live(&mut live_ids, *src2);
                }
            }

            Instruction::Select {
                dest,
                cond,
                if_true,
                if_false,
            } => {
                if live_ids[dest.0 as usize] {
                    live_instruction_count += 1;
                    mark_live(&mut live_ids, *cond);
                    mark_live(&mut live_ids, *if_true);
                    mark_live(&mut live_ids, *if_false);
                }
            }
        }
    }

    let mut instructions = Vec::with_capacity(live_instruction_count);

    for instr in graph.iter().filter(|it| match it.id() {
        Some(id) => live_ids[id.0 as usize],
        None => true,
    }) {
        instructions.push(instr.clone());
    }

    instructions
}

/// Moves register writes to be as close to their computations as possible.
pub fn register_writeback_shrinking(graph: &[Instruction]) -> Vec<Instruction> {
    let mut stores = vec![];
    let mut instrs = Vec::with_capacity(graph.len());

    for instr in graph {
        match instr {
            Instruction::WriteReg {
                dest,
                src: Source::Id(id),
                base,
            } => {
                stores.push((*dest, *id, *base));
            }

            _ => instrs.push(instr.clone()),
        }
    }

    for idx in (0..instrs.len()).rev() {
        if let Some(store_idx) = instrs[idx]
            .id()
            .and_then(|id| stores.iter().copied().position(|(_, src, _)| src == id))
        {
            let store = stores.remove(store_idx);
            // insert _after_ the instruction we just looked at
            instrs.insert(
                idx + 1,
                Instruction::WriteReg {
                    dest: store.0,
                    src: Source::Id(store.1),
                    base: store.2,
                },
            );
        }
    }

    instrs
}

#[cfg(test)]
mod test {
    use crate::ssa::lower;
    use crate::{instruction, opcode, register};
    use crate::{DisplayDeferSlice, Width};

    use insta::assert_display_snapshot;

    #[test]
    fn jal_basic_const_prop() {
        let ctx = lower::Context::new(0);

        let mut instrs = lower::terminal(
            ctx,
            instruction::Instruction::J(instruction::J {
                imm: 4096,
                rd: Some(register::RiscV::X4),
                opcode: opcode::J::JAL,
            }),
            4,
        );

        super::fold_and_prop_consts(&mut instrs);

        assert_display_snapshot!(DisplayDeferSlice(&instrs));
    }

    #[test]
    fn mem_read_write_all_opts() {
        let mut ctx = lower::Context::new(0);
        lower::non_terminal(
            &mut ctx,
            instruction::Instruction::IMem(instruction::IMem::new(
                0,
                Some(register::RiscV::X1),
                Some(register::RiscV::X2),
                opcode::IMem::LD(Width::DWord),
            )),
            4,
        );

        lower::non_terminal(
            &mut ctx,
            instruction::Instruction::I(instruction::I::new(
                100,
                Some(register::RiscV::X2),
                Some(register::RiscV::X2),
                opcode::I::ADDI,
            )),
            4,
        );

        lower::non_terminal(
            &mut ctx,
            instruction::Instruction::S(instruction::S::new(
                50,
                Some(register::RiscV::X2),
                Some(register::RiscV::X1),
                Width::DWord,
            )),
            4,
        );

        let mut instrs = lower::terminal(
            ctx,
            instruction::Instruction::Sys(instruction::Sys::new(opcode::RSys::EBREAK)),
            4,
        );

        super::fold_and_prop_consts(&mut instrs);
        let instrs = super::dead_instruction_elimination(&instrs);
        let instrs = super::register_writeback_shrinking(&instrs);

        assert_display_snapshot!(DisplayDeferSlice(&instrs));
    }

    #[test]
    fn max() {
        let mut instrs = crate::ssa::max_fn();

        super::fold_and_prop_consts(&mut instrs);
        let instrs = super::dead_instruction_elimination(&instrs);
        let instrs = super::register_writeback_shrinking(&instrs);

        assert_display_snapshot!(DisplayDeferSlice(&instrs));
    }

    #[test]
    fn jal_basic_die() {
        let ctx = lower::Context::new(0);

        let mut instrs = lower::terminal(
            ctx,
            instruction::Instruction::J(instruction::J {
                imm: 4096,
                rd: Some(register::RiscV::X4),
                opcode: opcode::J::JAL,
            }),
            4,
        );

        super::fold_and_prop_consts(&mut instrs);
        let instrs = super::dead_instruction_elimination(&instrs);

        assert_display_snapshot!(DisplayDeferSlice(&instrs));
    }
}
