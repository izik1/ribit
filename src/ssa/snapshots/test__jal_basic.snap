---
source: src/ssa/lower.rs
expression: DisplayDeferSlice(&instrs)
---
x4 = 4
ret 0, 4096
